// @sap-ux/preview-middleware/dist/client/cpe/quick-actions/quick-action-service.js
'use strict';
sap.ui.define([
    'sap/base/Log',
    'open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common',
    '../../i18n'
], function (Log, ___sap_ux_private_control_property_editor_common, ____i18n) {
    'use strict';
    const executeQuickAction = ___sap_ux_private_control_property_editor_common['executeQuickAction'];
    const quickActionListChanged = ___sap_ux_private_control_property_editor_common['quickActionListChanged'];
    const SIMPLE_QUICK_ACTION_KIND = ___sap_ux_private_control_property_editor_common['SIMPLE_QUICK_ACTION_KIND'];
    const NESTED_QUICK_ACTION_KIND = ___sap_ux_private_control_property_editor_common['NESTED_QUICK_ACTION_KIND'];
    const updateQuickAction = ___sap_ux_private_control_property_editor_common['updateQuickAction'];
    const getTextBundle = ____i18n['getTextBundle'];
    class QuickActionService {
        sendAction = () => {
        };
        actions = [];
        constructor(rta, outlineService, registries) {
            this.rta = rta;
            this.outlineService = outlineService;
            this.registries = registries;
        }
        async init(sendAction, subscribe) {
            this.sendAction = sendAction;
            this.actionService = await this.rta.getService('action');
            this.texts = await getTextBundle();
            subscribe(async action => {
                if (executeQuickAction.match(action)) {
                    const actionInstance = this.actions.filter(quickActionDefinition => quickActionDefinition.id === action.payload.id).pop();
                    if (!actionInstance) {
                        return;
                    }
                    const commands = await this.executeAction(actionInstance, action.payload);
                    for (const command of commands) {
                        await this.rta.getCommandStack().pushAndExecute(command);
                    }
                    if (actionInstance.forceRefreshAfterExecution) {
                        this.sendAction(updateQuickAction(actionInstance.getActionObject()));
                    }
                }
            });
            this.outlineService.onOutlineChange(async event => {
                await this.reloadQuickActions(event.detail.controlIndex);
            });
        }
        async reloadQuickActions(controlIndex) {
            const context = {
                controlIndex,
                manifest: this.rta.getRootControlInstance().getManifest(),
                actionService: this.actionService
            };
            const groups = [];
            for (const registry of this.registries) {
                for (const {title, definitions, view, key} of registry.getDefinitions(context)) {
                    const group = {
                        title,
                        actions: []
                    };
                    const actionContext = {
                        ...context,
                        view,
                        key,
                        rta: this.rta,
                        flexSettings: this.rta.getFlexSettings(),
                        resourceBundle: this.texts
                    };
                    for (const Definition of definitions) {
                        try {
                            const instance = new Definition(actionContext);
                            await instance.initialize();
                            this.addAction(group, instance);
                        } catch {
                            Log.warning(`Failed to initialize ${ Definition.name } quick action.`);
                        }
                    }
                    groups.push(group);
                }
            }
            this.sendAction(quickActionListChanged(groups));
        }
        addAction(group, instance) {
            if (instance.isActive) {
                const quickAction = instance.getActionObject();
                group.actions.push(quickAction);
                this.actions.push(instance);
            }
        }
        executeAction(actionInstance, payload) {
            if (payload.kind === SIMPLE_QUICK_ACTION_KIND && actionInstance.kind === SIMPLE_QUICK_ACTION_KIND) {
                return actionInstance.execute();
            }
            if (payload.kind === NESTED_QUICK_ACTION_KIND && actionInstance.kind === NESTED_QUICK_ACTION_KIND) {
                return actionInstance.execute(payload.path);
            }
            return Promise.resolve([]);
        }
    }
    var __exports = { __esModule: true };
    __exports.QuickActionService = QuickActionService;
    return __exports;
});
