// @sap-ux/adp-tooling/dist/writer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = generate;
exports.migrate = migrate;
const path_1 = require("path");
const mem_fs_1 = require("mem-fs");
const mem_fs_editor_1 = require("mem-fs-editor");
const options_1 = require("./options");
const project_utils_1 = require("./project-utils");
const tmplPath = (0, path_1.join)(__dirname, '../../templates/project');
/**
 * Set default values for optional properties.
 *
 * @param config configuration provided by the calling middleware
 * @returns enhanced configuration with default values
 */
function setDefaults(config) {
    const configWithDefaults = {
        app: { ...config.app },
        target: { ...config.target },
        ui5: { ...config.ui5 },
        deploy: config.deploy ? { ...config.deploy } : undefined,
        options: { ...config.options },
        flp: config.flp ? { ...config.flp } : undefined,
        customConfig: config.customConfig ? { ...config.customConfig } : undefined
    };
    configWithDefaults.app.title ??= `Adaptation of ${config.app.reference}`;
    configWithDefaults.app.layer ??= 'CUSTOMER_BASE';
    configWithDefaults.package ??= config.package ? { ...config.package } : {};
    configWithDefaults.package.name ??= config.app.id.toLowerCase().replace(/\./g, '-');
    configWithDefaults.package.description ??= configWithDefaults.app.title;
    if (configWithDefaults.flp && !configWithDefaults.flp.inboundId) {
        configWithDefaults.flp.addInboundId = true;
        configWithDefaults.flp.inboundId = `${configWithDefaults.app.id}.InboundID`;
    }
    return configWithDefaults;
}
/**
 * Writes the adp-project template to the mem-fs-editor instance.
 *
 * @param basePath - the base path
 * @param config - the writer configuration
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
async function generate(basePath, config, fs) {
    if (!fs) {
        fs = (0, mem_fs_editor_1.create)((0, mem_fs_1.create)());
    }
    const fullConfig = setDefaults(config);
    if (fullConfig.customConfig?.adp.environment === 'C' && fullConfig.flp) {
        (0, options_1.enhanceManifestChangeContentWithFlpConfig)(fullConfig.flp, fullConfig.app.id, fullConfig.app.content);
    }
    (0, project_utils_1.writeTemplateToFolder)((0, path_1.join)(tmplPath, '**/*.*'), (0, path_1.join)(basePath), fullConfig, fs);
    await (0, project_utils_1.writeUI5DeployYaml)(basePath, fullConfig, fs);
    await (0, project_utils_1.writeUI5Yaml)(basePath, fullConfig, fs);
    return fs;
}
/**
 * Writes the adp-project template to the mem-fs-editor instance during migration.
 *
 * @param basePath - the base path
 * @param config - the writer configuration
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
async function migrate(basePath, config, fs) {
    if (!fs) {
        fs = (0, mem_fs_editor_1.create)((0, mem_fs_1.create)());
    }
    const fullConfig = setDefaults(config);
    // Copy the specified files to target project
    fs.copyTpl((0, path_1.join)(tmplPath, '**/ui5.yaml'), (0, path_1.join)(basePath), fullConfig, undefined, {
        globOptions: { dot: true }
    });
    fs.copyTpl((0, path_1.join)(tmplPath, '**/package.json'), (0, path_1.join)(basePath), fullConfig, undefined, {
        globOptions: { dot: true }
    });
    fs.copyTpl((0, path_1.join)(tmplPath, '**/gitignore.tmpl'), (0, path_1.join)(basePath), fullConfig, undefined, {
        globOptions: { dot: true },
        processDestinationPath: (filePath) => filePath.replace(/gitignore.tmpl/g, '.gitignore')
    });
    // delete .che folder
    if (fs.exists((0, path_1.join)(basePath, '.che/project.json'))) {
        fs.delete((0, path_1.join)(basePath, '.che/*'));
    }
    // delete neo-app.json
    if (fs.exists((0, path_1.join)(basePath, 'neo-app.json'))) {
        fs.delete((0, path_1.join)(basePath, 'neo-app.json'));
    }
    await (0, project_utils_1.writeUI5Yaml)(basePath, fullConfig, fs);
    await (0, project_utils_1.writeUI5DeployYaml)(basePath, fullConfig, fs);
    return fs;
}
