// @sap-ux/preview-middleware/dist/client/adp/quick-actions/fe-v4/change-table-columns.js
'use strict';
sap.ui.define([
    'sap/ui/dt/OverlayUtil',
    'sap/ui/fl/apply/api/FlexRuntimeInfoAPI',
    'open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common',
    '../../../cpe/quick-actions/utils',
    '../../../utils/core'
], function (OverlayUtil, FlexRuntimeInfoAPI, ___sap_ux_private_control_property_editor_common, _____cpe_quick_actions_utils, _____utils_core) {
    'use strict';
    const NESTED_QUICK_ACTION_KIND = ___sap_ux_private_control_property_editor_common['NESTED_QUICK_ACTION_KIND'];
    const getRelevantControlFromActivePage = _____cpe_quick_actions_utils['getRelevantControlFromActivePage'];
    const getControlById = _____utils_core['getControlById'];
    const CHANGE_TABLE_COLUMNS = 'change-table-columns';
    const ACTION_ID = 'CTX_SETTINGS0';
    const CONTROL_TYPE = 'sap.ui.mdc.Table';
    class ChangeTableColumnsQuickAction {
        kind = NESTED_QUICK_ACTION_KIND;
        type = CHANGE_TABLE_COLUMNS;
        get id() {
            return `${ this.context.key }-${ this.type }`;
        }
        isActive = false;
        isClearButtonEnabled = false;
        children = [];
        tableMap = {};
        constructor(context) {
            this.context = context;
        }
        async initialize() {
            let index = 0;
            for (const smartTable of getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, [CONTROL_TYPE])) {
                const hasVariantManagement = FlexRuntimeInfoAPI.hasVariantManagement({ element: smartTable });
                if (!hasVariantManagement) {
                    continue;
                }
                const actions = await this.context.actionService.get(smartTable.getId());
                const changeColumnAction = actions.find(action => action.id === ACTION_ID);
                if (changeColumnAction) {
                    this.children.push({
                        label: `'${ smartTable.getHeader() }' table`,
                        children: []
                    });
                    this.tableMap[`${ this.children.length - 1 }`] = index;
                    index++;
                }
            }
            if (this.children.length > 0) {
                this.isActive = true;
            }
        }
        getActionObject() {
            return {
                kind: NESTED_QUICK_ACTION_KIND,
                id: this.id,
                enabled: this.isActive,
                title: this.context.resourceBundle.getText('V4_QUICK_ACTION_CHANGE_TABLE_COLUMNS'),
                children: this.children
            };
        }
        async execute(path) {
            const index = this.tableMap[path];
            const smartTables = getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, [CONTROL_TYPE]);
            for (let i = 0; i < smartTables.length; i++) {
                if (i === index) {
                    const section = getControlById(smartTables[i].getId());
                    const controlOverlay = OverlayUtil.getClosestOverlayFor(section);
                    if (controlOverlay) {
                        controlOverlay.setSelected(true);
                    }
                    const hasVariantManagement = FlexRuntimeInfoAPI.hasVariantManagement({ element: smartTables[i] });
                    if (!hasVariantManagement) {
                        continue;
                    }
                    await this.context.actionService.execute(smartTables[i].getId(), ACTION_ID);
                }
            }
            return [];
        }
    }
    var __exports = { __esModule: true };
    __exports.CHANGE_TABLE_COLUMNS = CHANGE_TABLE_COLUMNS;
    __exports.ChangeTableColumnsQuickAction = ChangeTableColumnsQuickAction;
    return __exports;
});
