// @sap-ux/preview-middleware/dist/client/adp/quick-actions/fe-v2/table-quick-action-base.js
'use strict';
sap.ui.define([
    'open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common',
    'sap/ui/dt/OverlayUtil',
    '../../../cpe/quick-actions/utils',
    '../../../utils/core',
    '../../../utils/version'
], function (___sap_ux_private_control_property_editor_common, OverlayUtil, _____cpe_quick_actions_utils, _____utils_core, _____utils_version) {
    'use strict';
    const NESTED_QUICK_ACTION_KIND = ___sap_ux_private_control_property_editor_common['NESTED_QUICK_ACTION_KIND'];
    const getParentContainer = _____cpe_quick_actions_utils['getParentContainer'];
    const getRelevantControlFromActivePage = _____cpe_quick_actions_utils['getRelevantControlFromActivePage'];
    const getControlById = _____utils_core['getControlById'];
    const isA = _____utils_core['isA'];
    const isManagedObject = _____utils_core['isManagedObject'];
    const getUi5Version = _____utils_version['getUi5Version'];
    const isLowerThanMinimalUi5Version = _____utils_version['isLowerThanMinimalUi5Version'];
    const SMART_TABLE_ACTION_ID = 'CTX_COMP_VARIANT_CONTENT';
    const M_TABLE_ACTION_ID = 'CTX_ADD_ELEMENTS_AS_CHILD';
    const SETTINGS_ID = 'CTX_SETTINGS';
    const ICON_TAB_BAR_TYPE = 'sap.m.IconTabBar';
    const SMART_TABLE_TYPE = 'sap.ui.comp.smarttable.SmartTable';
    const M_TABLE_TYPE = 'sap.m.Table';
    async function getActionId(table) {
        const {major, minor} = await getUi5Version();
        if (isA(SMART_TABLE_TYPE, table)) {
            if (major === 1 && minor === 96) {
                return [SETTINGS_ID];
            } else {
                return [SMART_TABLE_ACTION_ID];
            }
        }
        return [
            M_TABLE_ACTION_ID,
            SETTINGS_ID
        ];
    }
    class TableQuickActionDefinitionBase {
        kind = NESTED_QUICK_ACTION_KIND;
        get id() {
            return `${ this.context.key }-${ this.type }`;
        }
        isActive = false;
        children = [];
        tableMap = {};
        get textKey() {
            return this.defaultTextKey;
        }
        constructor(type, controlTypes, defaultTextKey, context, includeServiceAction) {
            this.type = type;
            this.controlTypes = controlTypes;
            this.defaultTextKey = defaultTextKey;
            this.context = context;
            this.includeServiceAction = includeServiceAction;
        }
        async initialize() {
            const version = await getUi5Version();
            if (isLowerThanMinimalUi5Version(version, {
                    major: 1,
                    minor: 96
                })) {
                this.isActive = false;
                return;
            }
            const iconTabBarfilterMap = this.buildIconTabBarFilterMap();
            for (const table of getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, this.controlTypes)) {
                const tabKey = Object.keys(iconTabBarfilterMap).find(key => table.getId().endsWith(key));
                const section = getParentContainer(table, 'sap.uxap.ObjectPageSection');
                if (section) {
                    this.collectChildrenInSection(section, table);
                } else if (this.iconTabBar && tabKey) {
                    this.children.push({
                        label: `'${ iconTabBarfilterMap[tabKey] }' table`,
                        children: []
                    });
                    this.tableMap[`${ this.children.length - 1 }`] = {
                        table,
                        iconTabBarFilterKey: tabKey,
                        tableUpdateEventAttachedOnce: false
                    };
                } else {
                    this.processTable(table);
                }
                if (this.includeServiceAction) {
                    const actions = await this.context.actionService.get(table.getId());
                    const actionsIds = await getActionId(table);
                    const changeColumnAction = actionsIds.find(actionId => actions.findIndex(action => action.id === actionId) > -1);
                    Object.keys(this.tableMap).forEach(key => {
                        this.tableMap[key].changeColumnActionId = changeColumnAction;
                    });
                }
            }
            if (this.children.length > 0) {
                this.isActive = true;
            }
        }
        getTableLabel(table) {
            if (isA(SMART_TABLE_TYPE, table)) {
                const header = table.getHeader();
                if (header) {
                    return `'${ header }' table`;
                }
            }
            if (isA(M_TABLE_TYPE, table)) {
                const tilte = table?.getHeaderToolbar()?.getTitleControl()?.getText();
                if (tilte) {
                    return `'${ tilte }' table`;
                }
            }
            return 'Unnamed table';
        }
        buildIconTabBarFilterMap() {
            const iconTabBarFilterMap = {};
            const tabBar = getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, [ICON_TAB_BAR_TYPE])[0];
            if (tabBar) {
                const control = getControlById(tabBar.getId());
                if (isA(ICON_TAB_BAR_TYPE, control)) {
                    this.iconTabBar = control;
                    for (const item of control.getItems()) {
                        if (isManagedObject(item) && isA('sap.m.IconTabFilter', item)) {
                            iconTabBarFilterMap[item.getKey()] = item.getText();
                        }
                    }
                }
            }
            return iconTabBarFilterMap;
        }
        collectChildrenInSection(section, table) {
            const layout = getParentContainer(table, 'sap.uxap.ObjectPageLayout');
            const subSections = section.getSubSections();
            const subSection = getParentContainer(table, 'sap.uxap.ObjectPageSubSection');
            if (subSection) {
                if (subSections?.length === 1) {
                    this.processTable(table, {
                        section,
                        subSection: subSections[0],
                        layout
                    });
                } else if (subSections.length > 1) {
                    const sectionChild = this.children.find(val => val.label === `${ section.getTitle() } section`);
                    let tableMapIndex = `${ this.children.length - 1 }`;
                    if (!sectionChild) {
                        tableMapIndex = `${ tableMapIndex }/0`;
                        this.children.push({
                            label: `'${ section?.getTitle() }' section`,
                            children: [{
                                    label: this.getTableLabel(table),
                                    children: []
                                }]
                        });
                    } else {
                        tableMapIndex = `${ tableMapIndex }/${ sectionChild.children.length - 1 }`;
                        sectionChild.children.push({
                            label: this.getTableLabel(table),
                            children: []
                        });
                    }
                    this.tableMap[tableMapIndex] = {
                        table,
                        sectionInfo: {
                            section,
                            subSection,
                            layout
                        },
                        tableUpdateEventAttachedOnce: false
                    };
                }
            }
        }
        processTable(table, sectionInfo) {
            if (isA(SMART_TABLE_TYPE, table) || isA('sap.ui.table.TreeTable', table)) {
                this.children.push({
                    label: this.getTableLabel(table),
                    children: []
                });
            }
            if (isA(M_TABLE_TYPE, table)) {
                this.children.push({
                    label: this.getTableLabel(table),
                    children: []
                });
            }
            this.tableMap[`${ this.children.length - 1 }`] = {
                table,
                sectionInfo: sectionInfo,
                tableUpdateEventAttachedOnce: false
            };
        }
        selectOverlay(table) {
            const controlOverlay = OverlayUtil.getClosestOverlayFor(table);
            if (controlOverlay) {
                controlOverlay.setSelected(true);
            }
        }
        getActionObject() {
            return {
                kind: NESTED_QUICK_ACTION_KIND,
                id: this.id,
                enabled: this.isActive,
                title: this.context.resourceBundle.getText(this.textKey),
                children: this.children
            };
        }
    }
    var __exports = { __esModule: true };
    __exports.TableQuickActionDefinitionBase = TableQuickActionDefinitionBase;
    return __exports;
});
