// @sap-ux/preview-middleware/dist/client/adp/quick-actions/simple-quick-action-base.js
'use strict';
sap.ui.define([
    'open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common',
    '../../cpe/quick-actions/utils'
], function (___sap_ux_private_control_property_editor_common, ____cpe_quick_actions_utils) {
    'use strict';
    const SIMPLE_QUICK_ACTION_KIND = ___sap_ux_private_control_property_editor_common['SIMPLE_QUICK_ACTION_KIND'];
    const getRelevantControlFromActivePage = ____cpe_quick_actions_utils['getRelevantControlFromActivePage'];
    class SimpleQuickActionDefinitionBase {
        kind = SIMPLE_QUICK_ACTION_KIND;
        get id() {
            return `${ this.context.key }-${ this.type }`;
        }
        get isActive() {
            return this.control !== undefined;
        }
        get textKey() {
            return this.defaultTextKey;
        }
        constructor(type, controlTypes, defaultTextKey, context) {
            this.type = type;
            this.controlTypes = controlTypes;
            this.defaultTextKey = defaultTextKey;
            this.context = context;
        }
        initialize() {
            for (const control of getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, this.controlTypes)) {
                this.control = control;
                break;
            }
        }
        getActionObject() {
            return {
                kind: SIMPLE_QUICK_ACTION_KIND,
                id: this.id,
                enabled: this.isActive,
                title: this.context.resourceBundle.getText(this.textKey)
            };
        }
    }
    var __exports = { __esModule: true };
    __exports.SimpleQuickActionDefinitionBase = SimpleQuickActionDefinitionBase;
    return __exports;
});
